Sources/GRPCInterceptors/Tracing/ClientOTelTracingInterceptor.swift

```diff
@@ -0,0 +1,196 @@
+    self.emitEventOnEachWrite = emitEventOnEachWrite
+  }
+
+  /// This interceptor will inject as the request's metadata whatever `ServiceContext` key-value pairs
+  /// have been made available by the tracing implementation bootstrapped in your application.
+  ///
+  /// Which key-value pairs are injected will depend on the specific tracing implementation
+  /// that has been configured when bootstrapping `swift-distributed-tracing` in your application.
+  ///
+  /// It will also inject all required and recommended span and event attributes, and set span status, as defined by OpenTelemetry's
+  /// documentation on:
+  /// - https://opentelemetry.io/docs/specs/semconv/rpc/rpc-spans
+  /// - https://opentelemetry.io/docs/specs/semconv/rpc/grpc/
+  public func intercept<Input, Output>(
+    request: StreamingClientRequest<Input>,
+    context: ClientContext,
+    next: (
+      StreamingClientRequest<Input>,
+      ClientContext
+    ) async throws -> StreamingClientResponse<Output>
+  ) async throws -> StreamingClientResponse<Output> where Input: Sendable, Output: Sendable {
+    var request = request
+    let tracer = InstrumentationSystem.tracer
+    let serviceContext = ServiceContext.current ?? .topLevel
+
+    tracer.inject(
+      serviceContext,
+      into: &request.metadata,
+      using: self.injector
+    )
+
+    return try await tracer.withSpan(
+      context.descriptor.fullyQualifiedMethod,
+      context: serviceContext,
+      ofKind: .client
+    ) { span in
+      self.setOTelSpanAttributes(into: span, context: context)
+
+      span.addEvent("Request started")
+
+      let wrappedProducer = request.producer
+      if self.emitEventOnEachWrite {
+        request.producer = { writer in
+          let messageSentCounter = Atomic(1)
+          let eventEmittingWriter = HookedWriter(
+            wrapping: writer,
+            beforeEachWrite: {},
+            afterEachWrite: {
+              var event = SpanEvent(name: "rpc.message")
+              event.attributes.rpc.messageType = "SENT"
+              event.attributes.rpc.messageID =
+                messageSentCounter
+                .wrappingAdd(1, ordering: .sequentiallyConsistent)
+                .oldValue
+              span.addEvent(event)
+            }
+          )
+          try await wrappedProducer(RPCWriter(wrapping: eventEmittingWriter))
+          span.addEvent("Request ended")
+        }
+      } else {
+        request.producer = { writer in
+          try await wrappedProducer(RPCWriter(wrapping: writer))
+          span.addEvent("Request ended")
+        }
+      }
+
+      var response = try await next(request, context)
+      switch response.accepted {
+      case .success(var success):
+        span.addEvent("Received response start")
+        span.attributes.rpc.grpcStatusCode = 0
+        if self.emitEventOnEachWrite {
+          let messageReceivedCounter = Atomic(1)
+          let onEachPartRecordingSequence = success.bodyParts.map { element in
```

Comment:
IIRC `element` could include the status here which we need to handle as well

Reply:
`element` is a `StreamingClientResponse.Contents.BodyPart`, which can be either a `message` or `trailingMetadata` - status isn't included, so no need to handle it.
However, the sequence can throw and we'll have to map the errors to span statuses, I did miss that so I'll change it.